lib: new error implementation

BridgeAR opened the pull request:
This implements a function based error system. Instead of passing in the
error code as first argument, the error code itself is a error class.
It already contains the correct error type, so while adding a new
error no one has to think about the error type anymore. In case a
single error code has more than one error type, the error class has
properties for the non default error types. Those can be used as
fallback.

This prevents typos, makes the implementation easier and it is less
verbose when writing the code for a new error.

The implementation itself does not interfere with the old
implementation. So the old and the new system can co-exist and it is
possible to slowly migrate the old ones to the new system.

By doing this I also realized that there are multiple error codes that
are currently not used at all. I also found other mistakes where a
wrong error type was in place.

So far this is only WIP and I wanted to get some feedback.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
lib

addaleax commented on lib/internal/errors.js:
```diff
@@ -86,6 +87,51 @@ function makeNodeError(Base) {
+
+    get name() {
+      return `${super.name} [${key}]`;
+    }
```
In this case we might also want to create a `static get name()` so that the class itself can be disambiguated from other classes of this type?

TimothyGu replied:
Or even better, `NodeError.name = ...`.

BridgeAR replied:
I thought about this a few times now and I decided not to do that out of the following reasons:

1) We only use the class internal and do not expose it directly.
2) The main class has sub classes as properties and those should make it easy to identify the correct one.
3) If we add the name, the original `NodeError` name will not be there and I think it should be.

devsnek commented on lib/internal/errors.js:
```diff
@@ -86,6 +87,51 @@ function makeNodeError(Base) {
+      return key;
+    }
+
+    set code(value) {
```
is there any use case for this?

mscdex replied:
My guess is that userland modules (not aware of the standardized node core error objects) may have already been setting this property to something else.

jasnell replied:
Yes, this was specifically asked for by userland module authors who need the ability to keep specifying their own codes.

jasnell commented on lib/console.js:
```diff
@@ -35,12 +35,12 @@ function Console(stdout, stderr, ignoreErrors = true) {
   }
   if (!stdout || typeof stdout.write !== 'function') {
-    throw new errors.TypeError('ERR_CONSOLE_WRITABLE_STREAM', 'stdout');
+    throw new ERR_CONSOLE_WRITABLE_STREAM('stdout');
```
The only challenge I really have with this is that it makes it far less obvious what *kind* of error is being thrown here.

BridgeAR replied:
That is also the charming side: you do not have to care about that as it was already taken care about at a earlier step. And we can verify much better that it actually has the correct type. Now you can see the types in the overview in the errors.js file. And I spotted multiple errors due to this.

BridgeAR commented:
I updated all error codes and removed the breaking change. It can now be used for all error codes. I also added quite a few comments about wrong error types. Please have a look. I personally suggest to keep the comments for now and change those in a different PR as that is breaking.

CI https://ci.nodejs.org/job/node-test-pull-request/13264/

joyeecheung commented on lib/internal/errors.js:
```diff
@@ -86,6 +87,54 @@ function makeNodeError(Base) {
   };
 }
 
+function makeNodeError2(Base, key) {
```
Can this be renamed to something more descriptive e.g. makeNodeErrorWithCode?

BridgeAR commented:
@mhdawson one NAPI code got removed that was not used (`ERR_NAPI_CONS_PROTOTYPE_OBJECT`). I checked every code that I removed and there are no hits anywhere for that code besides in the docs and errors.js. You can spot all removed error codes easily by looking at the documentation (I removed all of those from the documentation as well).

BridgeAR commented:
I changed the name. Running a Light-CI just to verify the minor changes since the full CI:

https://ci.nodejs.org/job/node-test-commit-light/307/console

BridgeAR commented:
Landed in 887a2ba92eec98ddf1e0953e742419236eff75d7...95bae858094450bd01f858ddd358ae0f83865157

MylesBorins commented:
Should this be backported to `v9.x-staging`? If yes please follow [the guide](https://github.com/nodejs/node/blob/master/doc/guides/backporting-to-release-lines.md) and raise a backport PR, if not let me know or add the `dont-land-on` label.

BridgeAR commented:
Backported in #19244

MylesBorins commented:
Backport requested for 8.x in https://github.com/nodejs/node/pull/19244
